p2p/protocol/autonatv2/server.go

```diff
@@ -0,0 +1,394 @@
+		}
+		remain -= len(msg.GetDialDataResponse().Data)
+		// Check if the peer is not sending too little data forcing us to just do a lot of compute
+		if len(msg.GetDialDataResponse().Data) < 100 && remain > 0 {
+			return fmt.Errorf("dial data msg too small: %d", len(msg.GetDialDataResponse().Data))
+		}
+	}
+	return nil
+}
+
+func (as *server) dialBack(p peer.ID, addr ma.Multiaddr, nonce uint64) pb.DialStatus {
+	ctx, cancel := context.WithTimeout(context.Background(), dialBackDialTimeout)
+	ctx = network.WithForceDirectDial(ctx, "autonatv2")
+	as.dialerHost.Peerstore().AddAddr(p, addr, peerstore.TempAddrTTL)
+	defer func() {
+		cancel()
+		as.dialerHost.Network().ClosePeer(p)
+		as.dialerHost.Peerstore().ClearAddrs(p)
+		as.dialerHost.Peerstore().RemovePeer(p)
+	}()
+
+	err := as.dialerHost.Connect(ctx, peer.AddrInfo{ID: p})
+	if err != nil {
+		return pb.DialStatus_E_DIAL_ERROR
+	}
+
+	s, err := as.dialerHost.NewStream(ctx, p, DialBackProtocol)
+	if err != nil {
+		return pb.DialStatus_E_DIAL_BACK_ERROR
+	}
+
+	defer s.Close()
+	s.SetDeadline(as.now().Add(dialBackStreamTimeout))
+
+	w := pbio.NewDelimitedWriter(s)
+	if err := w.WriteMsg(&pb.DialBack{Nonce: nonce}); err != nil {
+		s.Reset()
+		return pb.DialStatus_E_DIAL_BACK_ERROR
+	}
+
+	// Since the underlying connection is on a separate dialer, it'll be closed after this
+	// function returns. Connection close will drop all the queued writes. To ensure message
+	// delivery, do a CloseWrite and read a byte from the stream. The peer actually sends a
+	// response of type DialBackResponse but we only care about the fact that the DialBack
+	// message has reached the peer. So we ignore that message on the read side.
+	s.CloseWrite()
+	s.SetDeadline(as.now().Add(5 * time.Second)) // 5 is a magic number
+	b := make([]byte, 1)                         // Read 1 byte here because 0 len reads are free to return (0, nil) immediately
+	s.Read(b)
+
+	return pb.DialStatus_OK
+}
+
+// rateLimiter implements a sliding window rate limit of requests per minute. It allows 1 concurrent request
+// per peer. It rate limits requests globally, at a peer level and depending on whether it requires dial data.
+type rateLimiter struct {
+	// PerPeerRPM is the rate limit per peer
+	PerPeerRPM int
+	// RPM is the global rate limit
+	RPM int
+	// DialDataRPM is the rate limit for requests that require dial data
+	DialDataRPM int
+
+	mu           sync.Mutex
+	reqs         []entry
+	peerReqs     map[peer.ID][]time.Time
+	dialDataReqs []time.Time
+	// ongoingReqs tracks in progress requests. This is used to disallow multiple concurrent requests by the
+	// same peer
+	// TODO: Should we allow a few concurrent requests per peer?
+	ongoingReqs map[peer.ID]struct{}
+
+	now func() time.Time // for tests
+}
+
+type entry struct {
+	PeerID peer.ID
+	Time   time.Time
+}
+
+func (r *rateLimiter) Accept(p peer.ID) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.peerReqs == nil {
```

Comment:
Should this happen after cleanup and the checks below? Maybe right before the assignments on L321.

Reply:
It's easier to reason if it's set upfront, right?

Reply:
No strong opinion. We'll only do this ever once so it's fine.